Checkout: Add credit card component to replace credit card list items

dechov opened the pull request:
New `components/credit-card` module.

- The `<CreditCard>` component can either render a stored card (if passed a `card` prop), or wrap its `children` (such as a new card form). Thus, it serves to replace the [`section`](https://github.com/Automattic/wp-calypso/blob/24902a2c50a6b894b42951dd15a65551516d0002/client/my-sites/checkout/checkout/credit-card-selector.jsx#L56-L71) method of the `<CreditCardSelector>` component in `my-sites/checkout`.

- `<StoredCard>` (as [used](https://github.com/Automattic/wp-calypso/blob/24902a2c50a6b894b42951dd15a65551516d0002/client/me/purchases/credit-cards/credit-card-delete.jsx#L17) in the billing history view in `me/purchases`), is moved into this component directory as a child, and can continue to be imported directly into the billing history view.

Will open another PR off of these changes, to import this component into `my-sites/checkout` and `me/purchases`. (Changes in this PR taken from original WIP branch in https://github.com/Automattic/wp-calypso/pull/22840.)

dechov commented:
@folletto Absolutely – grabbing screenshots from `update/credit-card-lists-to-use-credit-card-component`, which branches off of this and actually uses the components:

- 5 `<CreditCard>`s here in this list, the top one being `selected`, the top four receiving a `card`, and the bottom one receiving `children` instead:

<img width="680" alt="screen shot 2018-04-30 at 4 34 51 pm" src="https://user-images.githubusercontent.com/1867547/39449126-37ff82e8-4c95-11e8-8f4c-d75e6fbc7d19.png">

- 1 `<CreditCard>` here (directly visible as the outer border only), receiving `children`: a `<StoredCard>` and a "Delete" button:
<img width="675" alt="screen shot 2018-04-30 at 4 51 39 pm" src="https://user-images.githubusercontent.com/1867547/39449698-cf5e5410-4c96-11e8-9c7d-64c81fb47da6.png">

- `<StoredCard>`:
<img width="565" alt="screen shot 2018-04-30 at 4 35 30 pm" src="https://user-images.githubusercontent.com/1867547/39449132-3ae4c130-4c95-11e8-94aa-6176af73a18b.png">

Note: the outer border was previously implemented in duplicate fashion but was nonetheless visually consistent between checkout and billing pages, so it is the one thing that is strictly indicative of a `<CreditCard>`. The `<StoredCard>` was already used in both places, but – in the current implementation – is unnecessary if a `card` is all that's being rendered inside the `<CreditCard>`.

Let me know any feedback on this approach, or if anything is unclear. Thanks!

folletto commented:
Looks good to me. :)

Question: shouldn't the credit card type be properly capitalized? (i.e. "Visa", "Mastercard", ...)?

dechov commented:
@folletto Yup, this PR is almost purely a refactor of existing code, but [such a change](https://github.com/Automattic/wp-calypso/compare/update/enhance-credit-card-component-summary-and-responsiveness#diff-be276b2d44cd85bb42de8149145dce8eR13) will be in a subsequent PR.

gwwar commented on client/components/credit-card/stored-card.jsx:
```diff
@@ -0,0 +1,49 @@
+
+export const cardType = PropTypes.shape( {
+	card: PropTypes.string,
+	card_type: PropTypes.string,
```
We'll probably want to clean up the 🐍 case here later

dechov replied:
Done in https://github.com/Automattic/wp-calypso/pull/24485/commits/646bea9c3b0278b73c80ce0cf34253645391eb02, as noted below. Thanks!

gwwar commented on client/components/credit-card/README.md:
```diff
@@ -0,0 +1,4 @@
+CreditCard
+==============
+
+The `CreditCard` React component serves as a container for a selectable credit card list item, and can display a stored credit card (passed as `card` prop).
```
I won't block on this, but let's add more detail here. See https://github.com/Automattic/wp-calypso/blob/master/client/components/button/README.md for a good example, this populates usage docs in https://wpcalypso.wordpress.com/devdocs/design/button

gwwar commented on client/components/credit-card/stored-card.jsx:
```diff
@@ -0,0 +1,49 @@
+		const expirationDate = this.props.moment( card.expiry ).format( 'MM/YY' );
+		const cardClasses = classNames(
+			'credit-card__stored-card',
+			card.card_type && card.card_type.toLowerCase()
```
card isn't required, so maybe
```jsx
get( card, [ 'card_type' ], '' ).toLowerCase()
```

gwwar replied:
oh hmm on second reading see if the value matches an enum value and add a modifier class? for example `is-amex` or `is-visa`

dmsnell replied:
to further extend what @gwwar wrote - how about being super explicit so that someone coming around to this later knows what to reasonably expect?

```js
const cardType = card && card.card_type.toLocaleLowerCase();
classNames(
  'credit-card__stored-card',
  {
    'is-amex': cardType === 'amex',
    'is-mastercard': cardType === 'mastercard',
    'is-visa': cardType === 'visa'
  }
);
``

gwwar commented on client/components/credit-card/stored-card.jsx:
```diff
@@ -0,0 +1,49 @@
+
+	render() {
+		const card = this.props.card;
+		const expirationDate = this.props.moment( card.expiry ).format( 'MM/YY' );
```
I know this was existing, but out of curiosity is this the best way to approach date localization cc @Automattic/i18n ? I know other countries prefer different groupings. Though this example if probably fine for most.

ramonjd replied:
Hi! 

There's an [ISO standard](https://en.wikipedia.org/wiki/ISO/IEC_7813) of `MM(\/|\-)YY` for credit cards and the like. So I'd say we shouldn't need to make it available for translation, and can leave it as it is. Also, it matches the format the user must enter at purchase anyway. 

See: https://translate.wordpress.com/projects/wpcom/-all-translated/240095/

I can find any _country_-specific formats, though I'd hazard a guess a say that any varying formats (where they exist) would be specific to a card/institution.

gwwar commented on client/components/credit-card/stored-card.scss:
```diff
@@ -0,0 +1,46 @@
+	justify-content: space-between;
+	padding: 15px 15px 15px 95px;
+
+	.credit-card__stored-card-name {
```
Did we need to group this under the parent for the added CSS specificity? It's okay to have this as top level otherwise.

dechov replied:
Changed in https://github.com/Automattic/wp-calypso/pull/24485/commits/90d6d02d50eaba69bbb03bf175d04e20cf5f47cd

gwwar commented on client/components/credit-card/stored-card.scss:
```diff
@@ -0,0 +1,46 @@
+// Credit Card Payment Box
```
Let's add a `/** @format */` to the top so prettier can handle formatting

gwwar replied:
Did we need this comment btw? It should be somewhat explanatory from the component structure.

dechov replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/24485/commits/380a53f0c4c8f19aa18bd19e66ce6898f20182cb

gwwar commented on client/components/credit-card/stored-card.scss:
```diff
@@ -0,0 +1,46 @@
+	}
+}
+
+$cc-brands: 'amex', 'diners', 'discover', 'jcb', 'mastercard', 'unionpay', 'visa';
```
We have a finite number of these, I'd say it's be okay to list these out unless you feel very strongly about it strongly. This helps when performing searches in code. For example if I were to look up `.credit-card__stored-card.amex` I wouldn't get a result.

gwwar replied:
I don't have super strong feelings, but I tend to avoid concat type code to help readers search later on. I try to optimize for lookup/reading over LoC savings within reason, since folks will be reading more than writing. In this case I'm fine with either approach, so don't count this as a blocker at all.

gwwar commented on client/components/credit-card/index.jsx:
```diff
@@ -0,0 +1,50 @@
+
+	render() {
+		const { card, selected, onSelect, className, children } = this.props;
+		const classes = classNames( 'credit-card', className, { selected, selectable: onSelect } );
```
`is-selected` I think is preferable with our [naming standards for modifiers](https://github.com/Automattic/wp-calypso/blob/ea90a8af13d48bbdc774b31543c17b7cb3c6c146/docs/coding-guidelines/css.md#csssass-coding-guidelines), but likely we haven't been super consistent with it.

dechov replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/24485/commits/dcb4f7fe5e8a9b7cabcac73ef10dc1051e59022c

gwwar commented on client/components/credit-card/style.scss:
```diff
@@ -0,0 +1,23 @@
+    &.selected {
+        cursor: default;
+        border-left: 3px solid $alert-green;
+        background-color: #fafdf6;
```
Custom color here. @folletto any suggestions for a named color?

folletto replied:
I'd suggest to use a derivative color in cases like this: since the base is `$alert-green`, we can try with `lighten( $alert-green, 30 )` (I haven't run the repo to see if that matches, tweak the number as needed to get a closer match).

fditrapani commented:
@gwwar: It might be good for @taggon to take a look.

gwwar commented:
Thanks for the updates @dechov let's give @taggon a chance to look too.

gwwar commented:
Sounds like for next steps all that's left is adding an example page before 🚢

dechov commented:
Apologies for the delay – I've rebased, made a change to the `StoredCard` props, and added docs. I'm not sure how to make the example show up in devdocs – doesn't seem to be working for me.

dechov commented on client/components/credit-card/stored-card.jsx:
```diff
@@ -0,0 +1,49 @@
+
+class StoredCard extends React.Component {
+	static propTypes = {
+		card: cardType,
```
I've chosen (https://github.com/Automattic/wp-calypso/pull/24485/commits/c2be1b04d37572a8a84a90f41637e295918ad561) to split this object into separate props, each of which is required.

dechov commented on client/components/credit-card/index.jsx:
```diff
@@ -0,0 +1,53 @@
+			'is-selectable': onSelect,
+		} );
+
+		return onSelect ? (
```
Done in https://github.com/Automattic/wp-calypso/pull/24485/commits/a9c35a418263b8b1578070cc3093b2029d5abeff

dmsnell commented on client/components/credit-card/docs/example.jsx:
```diff
@@ -0,0 +1,44 @@
+		cards: [
+			{
+				card: '4242',
+				card_type: 'visa',
```
these really ought to be camelCase as we have standard in Calypso.

dechov replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/24485/commits/646bea9c3b0278b73c80ce0cf34253645391eb02

dmsnell commented on client/components/credit-card/docs/example.jsx:
```diff
@@ -0,0 +1,44 @@
+	render() {
+		return (
+			<Fragment>
+				{ this.state.cards.map( ( card, i ) => <CreditCard key={ i } card={ card } /> ) }
```
the `key` here isn't unique to the card and this is one place where the bug can bite us - if we somehow remove a card from the list React could end up not redrawing or rearranging the others.

let's use something unique to the cards like their card number

```js
{ this.state.cards.map( card => <CreditCard key={ card.card } card={ card } /> ) }
```

dechov replied:
Done as part of https://github.com/Automattic/wp-calypso/pull/24485/commits/9c240b9f6bb3dee7cb127b730bc77b60b6fb1453

dmsnell commented on client/components/credit-card/docs/example.jsx:
```diff
@@ -0,0 +1,44 @@
+		return (
+			<Fragment>
+				{ this.state.cards.map( ( card, i ) => <CreditCard key={ i } card={ card } /> ) }
+				<CreditCard>Add another credit card...</CreditCard>
```
Don't forget to `translate()` this!

dmsnell commented on client/components/credit-card/index.jsx:
```diff
@@ -0,0 +1,52 @@
+
+class CreditCard extends React.Component {
+	static propTypes = {
+		card: PropTypes.shape( StoredCard.propTypes ),
```
nice use of an imported type

dmsnell commented on client/components/credit-card/stored-card.jsx:
```diff
@@ -0,0 +1,45 @@
+class StoredCard extends React.Component {
+	static propTypes = {
+		card: PropTypes.string.isRequired,
+		card_type: PropTypes.string.isRequired,
```
see other notes about snake_case vs camelCase

dmsnell commented on client/components/credit-card/stored-card.jsx:
```diff
@@ -0,0 +1,45 @@
+
+	render() {
+		const { card, card_type, name, expiry } = this.props;
+		const expirationDate = this.props.moment( expiry ).format( 'MM/YY' );
```
while already discussed and established that this is a standard format for credit card expiration months it's not obvious to the future reader why we aren't using a locale-based format string.

can we add an explanatory comment here to prevent future changes that would break the behavior?

```js
// the use of `MM/YY` is standard across credit-card forms and should not be localized
const expirationDate = this.props…
```

dechov replied:
Done in https://github.com/Automattic/wp-calypso/pull/24485/commits/bdcbd88445f8a34648b8cc2c0b7c0c0d233f0b6d

dmsnell commented on client/components/credit-card/stored-card.jsx:
```diff
@@ -0,0 +1,45 @@
+		const expirationDate = this.props.moment( expiry ).format( 'MM/YY' );
+		const cardClasses = classNames(
+			'credit-card__stored-card',
+			card_type && card_type.toLowerCase()
```
I left a comment about this elsewhere but maybe it go buried. being explicit here helps because in this construction we have no expectation on what values should appear here.

dechov replied:
👍 Done in https://github.com/Automattic/wp-calypso/pull/24485/commits/646bea9c3b0278b73c80ce0cf34253645391eb02

dmsnell commented on client/components/credit-card/docs/example.jsx:
```diff
@@ -0,0 +1,44 @@
+	state = {
+		cards: [
+			{
+				card: '4242',
```
`card` seems like a fuzzy term. I _think_ we mean _credit card number_ or maybe _last four digits of a credit card number_ - how would you feel about making this name more specific?

`lastFour` or `cardNumber` or `cardId` or something that makes sense would be possibilities

dechov replied:
Good call – I was sticking to the format already being used, but no reason to formalize it in the view if it's imperfect – done in https://github.com/Automattic/wp-calypso/pull/24485/commits/22dda8e457d74c6541d16bb7583c706f625065cd

a8ci18n commented on client/components/credit-card/docs/example.jsx:
```diff
@@ -0,0 +1,52 @@
+				) ) }
+				<CreditCard>
+					<div style={ { padding: 20 } }>
+						{ this.props.translate( 'Add another credit card...' ) }
```
Hi! I've found a possible matching string that has already been translated **2** times:
`translate( 'Add another credit card' )` *ES Score: 17*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

dechov commented:
Addressed latest feedback – thanks @dmsnell for taking a look. Also, properly added example to devdocs now:

<img width="664" alt="screen shot 2018-06-20 at 1 37 07 pm" src="https://user-images.githubusercontent.com/1867547/41674902-24bce2bc-748f-11e8-80b4-6d943920dcdc.png">

alisterscott commented:
Looks like this one just needs a few more reviews and 👍 ?

southp commented on client/components/credit-card/stored-card.jsx:
```diff
@@ -0,0 +1,54 @@
+				<span className="credit-card__stored-card-expiration-date">
+					{ this.props.translate( 'Expires %(date)s', {
+						args: { date: expirationDate },
+						context: 'date is of the form MM/YY',
```
I know this string has been written in this way for years, so just wanted to note that in this case we should use `comment` instead of `context`. 

In short, `comment` is the detailed info that you want to tell translators. `context` is used when you want to translate the same string in a different way from the current translation. You can find more details under the "options" section here for more details: https://github.com/Automattic/i18n-calypso

Let's just keep it this way so the end users won't lose the translation.

southp commented on client/components/credit-card/docs/example.jsx:
```diff
@@ -0,0 +1,53 @@
+}
+
+const LocalizedCreditCards = localize( CreditCards );
+LocalizedCreditCards.displayName = 'CreditCard';
```
Since you are using ES6 class already, you can just put `displayName` as a static data property right in the class definition as
```
static displayName = 'CreditCard';
```
A example: https://github.com/Automattic/wp-calypso/blob/master/client/components/forms/form-password-input/index.jsx#L19

Also, I assume it should be `CreditCards` instead of the singular form?

With that, line 50 ~ 53 can then be simplified as just `export default localize( CreditCards )`.

dechov replied:
Thanks for testing!

This `displayName` is intended to appear as the title of the component in devdocs. Though the example contains multiple `<CreditCard>`s, the component is still singular, following the pattern of other examples: https://github.com/Automattic/wp-calypso/blob/34365c471f635ea24e90e7a3a976b52df1f9e3f3/client/components/button/docs/example.jsx#L20-L21

And since `localize` transforms the `displayName`, I set it on the localized component rather than the original. Open to other ways of doing this, but I hadn't come across any.

gwwar commented:
Don't forget to land this one ✨
